kevm-pyk/src/kevm_pyk/utils.py

```diff
@@ -102,22 +140,58 @@ def run_prover(
     counterexample_info: bool = False,
     always_check_subsumption: bool = False,
     fast_check_subsumption: bool = False,
+    max_frontier_parallel: int = 1,
 ) -> bool:
     prover: APRProver | ImpliesProver
     try:
         if type(proof) is APRProof:
-            prover = APRProver(
-                kcfg_explore,
-                execute_depth=max_depth,
-                terminal_rules=terminal_rules,
-                cut_point_rules=cut_point_rules,
-                counterexample_info=counterexample_info,
-                always_check_subsumption=always_check_subsumption,
-                fast_check_subsumption=fast_check_subsumption,
-            )
-            prover.advance_proof(proof, max_iterations=max_iterations, fail_fast=fail_fast)
+
+            def create_prover() -> APRProver:
```

Comment:
As a next steps, you can move these initialization steps into the respective strategy classes. Then you can extract a function that picks a strategy for you:

```python3
if create_kcfg_explore is not None:
    return MultithreadedProvingStrategy(...) max_frontier_parallel > 1 else SequentialProvingStrategy(...)
elif kcfg_explore is not None:
    return SequentialProvingStrategy(...)
else:
   raise ValueError(...)
```

You can make `execute_depth`, etc. parameters of `__init__` on the abstract class. Cosider extracting a class to hold these parameters.

Reply:
I improved it by pushing the initialization steps into the strategies.